packages/flutter/lib/src/foundation/memory_allocations.dart

```diff
@@ -0,0 +1,284 @@
+
+  /// List of listeners.
+  ///
+  /// The elements are nullable, because the listeners should be removable
+  /// while iterating through the list.
+  List<ObjectEventListener?>? _listeners;
+
+  /// Register a listener that is called every time an object event is
+  /// dispatched.
+  ///
+  /// Listeners can be removed with [removeListener].
+  void addListener(ObjectEventListener listener){
+    if (_listeners == null) {
+      _listeners = <ObjectEventListener?>[];
+      _subscribeToSdkObjects();
+    }
+    _listeners!.add(listener);
+  }
+
+  /// Number of active notification loops.
+  ///
+  /// When equal to zero, we can delete listeners from the list,
+  /// otherwise should null them.
+  int _activeDispatchLoops = 0;
+
+  /// If true, listeners were nulled by [removeListener].
+  bool _listenersContainNulls = false;
+
+  /// Stop calling the given listener every time an object event is
+  /// dispatched.
+  ///
+  /// Listeners can be added with [addListener].
+  void removeListener(ObjectEventListener listener){
+    final List<ObjectEventListener?>? listeners = _listeners;
+    if (listeners == null) {
+      return;
+    }
+
+    if (_activeDispatchLoops > 0) {
+      // If there are active dispatch loops, listeners.remove
+      // should not be invoked, as it will
+      // break the dispatch loops correctness.
+      for (int i = 0; i < listeners.length; i++) {
+        if (listeners[i] == listener) {
+          listeners[i] = null;
+          _listenersContainNulls = true;
+        }
+      }
+    } else {
+      listeners.removeWhere((ObjectEventListener? l) => l == listener);
+      _checkListenersForEmptiness();
+    }
+  }
+
+  void _tryDefragmentListeners() {
+    if (_activeDispatchLoops > 0 || !_listenersContainNulls) {
+      return;
+    }
+    _listeners?.removeWhere((ObjectEventListener? e) => e == null);
+    _listenersContainNulls = false;
+    _checkListenersForEmptiness();
+  }
+
+  void _checkListenersForEmptiness() {
+    if (_listeners?.isEmpty ?? false) {
+        _listeners = null;
+        _unSubscribeFromSdkObjects();
+    }
+  }
+
+  /// Return true if there are listeners.
+  ///
+  /// If there is no listeners, the app can save on creating the event object.
+  bool get hasListeners {
+    if (_listenersContainNulls) {
+      return _listeners?.firstWhere((ObjectEventListener? l) => l != null) != null;
+    }
+    return _listeners?.isNotEmpty ?? false;
+  }
+
+  /// Dispatch a new object event to listeners.
+  ///
+  /// Exceptions thrown by listeners will be caught and reported using
+  /// [FlutterError.reportError].
+  ///
+  /// Listeners added during an event dispatching, will start being invoked
+  /// for next events, but will be skipped for this event.
+  ///
+  /// Listeners, removed during an event dispatching, will not be invoked
+  /// after the removal.
+  void dispatchObjectEvent(ObjectEventBuilder objectEventBuilder) {
+    final List<ObjectEventListener?>? listeners = _listeners;
```

Comment:
assert that kFlutterMemoryAllocationsEnabled is true?

And document that this must not be called if kFlutterMemoryAllocationsEnabled is false.

Reply:
I wonder if we should guard other API as well. Should you be able add/remove/check listeners when this is not enabled? My feeling is that you probably shouldn't since that will prevent the MemoryAllocation code from tree-shaking...

Reply:
hm, assert will not help, because kFlutterMemoryAllocationsEnabled is true in debug mode and asserts do not work in release mode.

Added description to this class explaining to use the class with the condition.

I think have some level of safety here because we will never advertise direct usage of this class, and the package leak_tracker will make sure to verify kFlutterMemoryAllocationsEnabled before adding listeners.

What do you think?

Reply:
Sounds good. Can you add the note about only calling this when `kFlutterMemoryAllocationsEnabled` is true to the method doc as well?